src/main/scala/scorex/core/network/NodeViewSynchronizer.scala

```diff
@@ -226,53 +257,78 @@ class NodeViewSynchronizer[TX <: Transaction,
       log.info(s"Got modifiers of type $typeId from remote connected peer: $remote")
       log.trace(s"Received modifier ids ${data._2.keySet.map(encoder.encode).mkString(",")}")
 
-      modifiers.foreach { case (id, _) => deliveryTracker.onReceive(typeId, id, remote) }
-
-      val (spam, fm) = modifiers.partition { case (id, _) => deliveryTracker.isSpam(id) }
+      val (fm, spam) = modifiers.partition { case (id, _) =>
+        deliveryTracker.onReceive(typeId, id, remote)
+      }
 
       if (spam.nonEmpty) {
         log.info(s"Spam attempt: peer $remote has sent a non-requested modifiers of type $typeId with ids" +
           s": ${spam.keys.map(encoder.encode)}")
         penalizeSpammingPeer(remote)
-        val mids = spam.keys.toSeq
-        deliveryTracker.deleteSpam(mids)
-      }
-
-      if (fm.nonEmpty) {
-        val mods = fm.values.toSeq
-        viewHolderRef ! ModifiersFromRemote(remote, typeId, mods)
       }
-  }
 
-  //local node sending object ids to remote
-  protected def requestFromLocal: Receive = {
-    case RequestFromLocal(peer, modifierTypeId, modifierIds) =>
-
-      if (modifierIds.nonEmpty) {
-        val msg = Message(requestModifierSpec, Right(modifierTypeId -> modifierIds), None)
-        peer.handlerRef ! msg
+      modifierSerializers.get(typeId) match {
+        case Some(companion) =>
+          fm.foreach { case (id, bytes) =>
+            companion.parseBytes(bytes) match {
+              case Success(mod) if !(id sameElements mod.id) =>
+                log.warn(s"Declared id ${encoder.encode(id)} is not equals to calculated one ${mod.encodedId}")
+                penalizeMisbehavingPeer(remote)
+                deliveryTracker.toUnknown(id)
+                None
+              case Failure(e) =>
+                log.warn(s"Failed to parse modifier ${encoder.encode(id)}", e)
+                penalizeMisbehavingPeer(remote)
+                deliveryTracker.toUnknown(id)
+                None
+              case Success(tx: TX@unchecked) if tx.modifierTypeId == Transaction.ModifierTypeId =>
+                viewHolderRef ! LocallyGeneratedTransaction[TX](tx)
+              case Success(pmod: PMOD@unchecked) =>
+                if (modifiersCache.contains(pmod.id) || historyReaderOpt.exists(_.contains(pmod))) {
```

Comment:
maybe it worth moving body of this case branch to a separate method to make `modifiersFromRemote` handler shorter